Multiple batch queues

gfr10598 opened the pull request:
This will be used primarily in production, to feed the etl-ndt-batch-parser.  Currently other appengine configs do not configure a batch pipeline.

Also removes the unused pull queue.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/etl/292)
<!-- Reviewable:end -->

stephen-soltesz commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[appengine/queue.yaml, line 24 at r2](https://reviewable.io:443/reviews/m-lab/etl/292#-KxPHMU668uMfP-t7mvF:-KxPHMU76cgbMJRUw0pL:bbjbeg0) ([raw file](https://github.com/m-lab/etl/blob/fd0ef1369c5327d0f4bef67496e970f7714222e1/appengine/queue.yaml#L24)):*
> ```YAML
>   rate: 0.8/s
>   bucket_size: 10
>   max_concurrent_requests: 100
> ```

We discussed in person that the day-of-week queues are for archives from that day.

I understand instance as "parser VM". And I understand 12 workers as the number of simultaneous archives parsed by an instance. 50 * 12 == 600. 

I don't understand the relationship between max_concurrent_requests (700) across all queues and 600 available workers. Why not make these equal?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/292)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[appengine/queue.yaml, line 24 at r2](https://reviewable.io:443/reviews/m-lab/etl/292#-KxPHMU668uMfP-t7mvF:-KxPiWaOrGbOF2hBiifm:bn6ygvu) ([raw file](https://github.com/m-lab/etl/blob/fd0ef1369c5327d0f4bef67496e970f7714222e1/appengine/queue.yaml#L24)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

We discussed in person that the day-of-week queues are for archives from that day.

I understand instance as "parser VM". And I understand 12 workers as the number of simultaneous archives parsed by an instance. 50 * 12 == 600. 

I don't understand the relationship between max_concurrent_requests (700) across all queues and 600 available workers. Why not make these equal?
</blockquote></details>

Fixed to use 560, and fixed comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/292)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[appengine/queue.yaml, line 17 at r3](https://reviewable.io:443/reviews/m-lab/etl/292#-KxTlC1vYh7JLcPMJwTD:-KxTlC1w2Hnl4m8NslR9:b-ymeyrt) ([raw file](https://github.com/m-lab/etl/blob/4672494c50880b9d9769237d5b2b566162d2ddd2/appengine/queue.yaml#L17)):*
> ```YAML
> 
> # BigQuery has a streaming insert quota per table, which we seem to hit
> # at around 3.0 tasks/second.  When reprocessing data older than 30 days,
> ```

It's still unclear to me how all the pieces fit together. I imagine there is something I'm missing from the big picture that I would appreciate your help understanding.

The BQ limit of 3 tasks/sec per table sounds like what we're designing around. Is this right?

It is unclear to me how placing historical archives into these queues accomplishes that.

* will historical archives be processed one month at a time? Or multiple months at once?
* tables use templates that are per-YYYYMMDD, correct? Why is there an advantage to funneling the archives through day-of-week queues?
* If all archives from a given date pass through a single queue, with up to 80 concurrent requests, all potentially to the same table, why won't the bq rate limit be exceeded?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/292)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[appengine/queue.yaml, line 24 at r2](https://reviewable.io:443/reviews/m-lab/etl/292#-KxPHMU668uMfP-t7mvF:-KxhGIeFrXc_M-_uga7s:b-83330m) ([raw file](https://github.com/m-lab/etl/blob/fd0ef1369c5327d0f4bef67496e970f7714222e1/appengine/queue.yaml#L24)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Fixed to use 560, and fixed comment.
</blockquote></details>

PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/292)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[appengine/queue.yaml, line 17 at r3](https://reviewable.io:443/reviews/m-lab/etl/292#-KxTlC1vYh7JLcPMJwTD:-KxhGTqQ8mDdU-LIsnaZ:bid09fl) ([raw file](https://github.com/m-lab/etl/blob/4672494c50880b9d9769237d5b2b566162d2ddd2/appengine/queue.yaml#L17)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

It's still unclear to me how all the pieces fit together. I imagine there is something I'm missing from the big picture that I would appreciate your help understanding.

The BQ limit of 3 tasks/sec per table sounds like what we're designing around. Is this right?

It is unclear to me how placing historical archives into these queues accomplishes that.

* will historical archives be processed one month at a time? Or multiple months at once?
* tables use templates that are per-YYYYMMDD, correct? Why is there an advantage to funneling the archives through day-of-week queues?
* If all archives from a given date pass through a single queue, with up to 80 concurrent requests, all potentially to the same table, why won't the bq rate limit be exceeded?
</blockquote></details>

We probably will process one month to three months at a time.
The advantage comes because we can have the cumulative throughput for the set of queues be quite high (7 tasks/sec), while the rate for any receiving table is still limited (to 1 task/sec).
Requests take on the order of 40 seconds to process, so the 80 concurrent tasks translates to about 2 tasks/sec.  But the 1.0 requests/sec will be the limiting factor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/292)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[appengine/queue.yaml, line 17 at r3](https://reviewable.io:443/reviews/m-lab/etl/292#-KxTlC1vYh7JLcPMJwTD:-KxhHAAZ7XIG6rxlfjpu:bij1ipm) ([raw file](https://github.com/m-lab/etl/blob/4672494c50880b9d9769237d5b2b566162d2ddd2/appengine/queue.yaml#L17)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

We probably will process one month to three months at a time.
The advantage comes because we can have the cumulative throughput for the set of queues be quite high (7 tasks/sec), while the rate for any receiving table is still limited (to 1 task/sec).
Requests take on the order of 40 seconds to process, so the 80 concurrent tasks translates to about 2 tasks/sec.  But the 1.0 requests/sec will be the limiting factor.
</blockquote></details>

Also, while the BQ limit is around 3 tasks/sec, I'm limiting each table to much lower number of 1 task/sec.  We can raise this in the future if we choose to.  I also now have logic in the inserter to retry failed inserts, so there is an additional rate limiting there, but at the expense of a lot of repeated json rendering and large outgoing HTTP requests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/292)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
I don't want to block you, but this is not easy for me to reason about yet.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/292#-:-Kxhn5esLr09f7h2cu1Q:b-6bvu8i)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[appengine/queue.yaml, line 17 at r3](https://reviewable.io:443/reviews/m-lab/etl/292#-KxTlC1vYh7JLcPMJwTD:-KxiNweoapj_UJ5mfpzR:bp7it2q) ([raw file](https://github.com/m-lab/etl/blob/4672494c50880b9d9769237d5b2b566162d2ddd2/appengine/queue.yaml#L17)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Also, while the BQ limit is around 3 tasks/sec, I'm limiting each table to much lower number of 1 task/sec.  We can raise this in the future if we choose to.  I also now have logic in the inserter to retry failed inserts, so there is an additional rate limiting there, but at the expense of a lot of repeated json rendering and large outgoing HTTP requests.
</blockquote></details>

Does the intuition gained from archive sizes created by scraper apply to historical archives crated by the old dropbox collector? Do these size & rate estimates still apply?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/292)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Thank you. This is much clearer and helpful for understanding the rationale such that someone else could pick it up to tune it. Added one comment that may or may not be helpful for making the numbers match.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[appengine/queue.yaml, line 48 at r5](https://reviewable.io:443/reviews/m-lab/etl/292#-KxiisDZ9vtggmhYLJGS:-KxiisD_K-Qtkybe_zCB:b6dfj86) ([raw file](https://github.com/m-lab/etl/blob/a57592fb7e26857159ad7b690aa5c6cc3fb89d01/appengine/queue.yaml#L48)):*
> ```YAML
> # of the very large older archives. But by limiting the rate based on
> # number of concurrent tasks, we can regulate the throughput independent
> # of the size of the archive files.  100 rows/second, divided by 1.3 rows/sec
> ```

If 60 works or is just an even more conservative number, this is no problem. Just observing that 100 / 1.3 == ~76.

---

*[appengine/queue.yaml, line 74 at r5](https://reviewable.io:443/reviews/m-lab/etl/292#-KxijfA2rVmALj1Y7iuY:-KxijfA2rVmALj1Y7iuZ:b-9pdp29) ([raw file](https://github.com/m-lab/etl/blob/a57592fb7e26857159ad7b690aa5c6cc3fb89d01/appengine/queue.yaml#L74)):*
> ```YAML
> # 1. Each queue must limit concurrent requests to about 60 to limit
> #    rate into individual tables to around 100 rows/second.
> # 2. We have 5 queues, to allow up to about 500 rows/second, to
> ```

5 * 60 * 1.3 = 390 -- should this be closer to 500 or have I misunderstood the numbers?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/292#-:-Kxij6GLGTQkMLxs0hRO:b-egf8gx)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[appengine/queue.yaml, line 48 at r5](https://reviewable.io:443/reviews/m-lab/etl/292#-KxiisDZ9vtggmhYLJGS:-KxiomOhNdBY-7zC-Fw9:bz99brz) ([raw file](https://github.com/m-lab/etl/blob/a57592fb7e26857159ad7b690aa5c6cc3fb89d01/appengine/queue.yaml#L48)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

If 60 works or is just an even more conservative number, this is no problem. Just observing that 100 / 1.3 == ~76.
</blockquote></details>

Tx.  That was a math in my head error.

---

*[appengine/queue.yaml, line 74 at r5](https://reviewable.io:443/reviews/m-lab/etl/292#-KxijfA2rVmALj1Y7iuY:-Kxiot08pb9cL0uWrce7:b-itgraf) ([raw file](https://github.com/m-lab/etl/blob/a57592fb7e26857159ad7b690aa5c6cc3fb89d01/appengine/queue.yaml#L74)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

5 * 60 * 1.3 = 390 -- should this be closer to 500 or have I misunderstood the numbers?
</blockquote></details>

Yup.  Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/292)*
<!-- Sent from Reviewable.io -->
